pkg/agent/puller.go

```diff
@@ -146,60 +146,72 @@ func (p *Puller) modelProcessor(modelName string, ops <-chan *ModelOp) {
 	// Load --> Unload = 0 (cancel first load)
 	// Load --> Unload --> Load = 1 Load (cancel second load?)
 	for modelOp := range ops {
-		switch modelOp.Op {
-		case Add:
-			p.logger.Infof("Downloading model from %s", modelOp.Spec.StorageURI)
-			err := p.Downloader.DownloadModel(modelName, modelOp.Spec)
-			if err != nil {
-				// If there is an error, we will NOT send a request. As such, to know about errors, you will
-				// need to call the error endpoint of the puller
-				p.logger.Errorf("Failed to download model %s with err %v", modelName, err)
-			} else {
-				// Load the model onto the model server
-				resp, err := http.Post(fmt.Sprintf("http://localhost:8080/v2/repository/models/%s/load", modelName),
-					"application/json",
-					bytes.NewBufferString("{}"))
+		func() {
```

Comment:
Nice catch here. 

How about extracting the function from the loop and using the occasion to further reduce indentation? Something like this maybe:

```diff
diff --git a/pkg/agent/puller.go b/pkg/agent/puller.go
index 683249ed..35f748fc 100644
--- a/pkg/agent/puller.go
+++ b/pkg/agent/puller.go
@@ -145,73 +145,74 @@ func (p *Puller) modelProcessor(modelName string, ops <-chan *ModelOp) {
 	// this is important for handling Load --> Unload requests sent in tandem
 	// Load --> Unload = 0 (cancel first load)
 	// Load --> Unload --> Load = 1 Load (cancel second load?)
-	for modelOp := range ops {
-		func() {
-			switch modelOp.Op {
-			case Add:
-				p.logger.Infof("Downloading model from %s", modelOp.Spec.StorageURI)
-				err := p.Downloader.DownloadModel(modelName, modelOp.Spec)
-				if err != nil {
-					// If there is an error, we will NOT send a request. As such, to know about errors, you will
-					// need to call the error endpoint of the puller
-					p.logger.Errorf("Failed to download model %s with err %v", modelName, err)
-				} else {
-					// Load the model onto the model server
-					resp, err := http.Post(fmt.Sprintf("http://localhost:8080/v2/repository/models/%s/load", modelName),
-						"application/json",
-						bytes.NewBufferString("{}"))
-					if err != nil {
-						// handle error
-						p.logger.Errorf("Failed to Load model %s", modelName)
-					} else {
-						defer func(Body io.ReadCloser) {
-							newErr := Body.Close()
-							if newErr != nil {
-								p.logger.Error(newErr, "failed to close body")
-							}
-						}(resp.Body)
-						if resp.StatusCode == 200 {
-							p.logger.Infof("Successfully loaded model %s", modelName)
-						} else {
-							body, err := io.ReadAll(resp.Body)
-							if err == nil {
-								p.logger.Infof("Failed to load model %s with status [%d] and resp:%v", modelName, resp.StatusCode, body)
-							}
-						}
-					}
+	var processOp = func(op *ModelOp) {
+		switch op.Op {
+		case Add:
+			p.logger.Infof("Downloading model from %s", op.Spec.StorageURI)
+			err := p.Downloader.DownloadModel(modelName, op.Spec)
+			if err != nil {
+				// If there is an error, we will NOT send a request. As such, to know about errors, you will
+				// need to call the error endpoint of the puller
+				p.logger.Errorf("Failed to download model %s with err %v", modelName, err)
+				return
+			}
+			// Load the model onto the model server
+			resp, err := http.Post(fmt.Sprintf("http://localhost:8080/v2/repository/models/%s/load", modelName),
+				"application/json",
+				bytes.NewBufferString("{}"))
+			if err != nil {
+				p.logger.Errorf("Failed to Load model %s", modelName)
+				return
+			}
+			defer func(body io.ReadCloser) {
+				newErr := body.Close()
+				if newErr != nil {
+					p.logger.Error(newErr, "failed to close body")
 				}
-			case Remove:
-				p.logger.Infof("unloading model %s", modelName)
-				// If there is an error, we will NOT do a delete... that could be problematic
-				if err := storage.RemoveDir(filepath.Join(p.Downloader.ModelDir, modelName)); err != nil {
-					p.logger.Error(err, "failing to delete model directory")
-				} else {
-					// unload model from model server
-					resp, err := http.Post(fmt.Sprintf("http://localhost:8080/v2/repository/models/%s/unload", modelName),
-						"application/json",
-						bytes.NewBufferString("{}"))
-					if err != nil {
-						// handle error
-						p.logger.Errorf("Failed to Unload model %s", modelName)
-					} else {
-						defer func(Body io.ReadCloser) {
-							newErr := Body.Close()
-							if newErr != nil {
-								p.logger.Error(newErr, "failed to close body")
-							}
-						}(resp.Body)
-						if resp.StatusCode == 200 {
-							p.logger.Infof("Successfully unloaded model %s", modelName)
-						} else {
-							body, err := io.ReadAll(resp.Body)
-							if err == nil {
-								p.logger.Infof("Failed to unload model %s with status [%d] and resp:%v", modelName, resp.StatusCode, body)
-							}
-						}
-					}
+			}(resp.Body)
+			if resp.StatusCode == 200 {
+				p.logger.Infof("Successfully loaded model %s", modelName)
+			} else {
+				body, err := io.ReadAll(resp.Body)
+				if err == nil {
+					p.logger.Infof("Failed to load model %s with status [%d] and resp:%v", modelName, resp.StatusCode, body)
 				}
 			}
-			p.completions <- modelOp
-		}()
+		case Remove:
+			p.logger.Infof("unloading model %s", modelName)
+			// If there is an error, we will NOT do a delete... that could be problematic
+			if err := storage.RemoveDir(filepath.Join(p.Downloader.ModelDir, modelName)); err != nil {
+				p.logger.Error(err, "failing to delete model directory")
+				return
+			}
+			// unload model from model server
+			resp, err := http.Post(fmt.Sprintf("http://localhost:8080/v2/repository/models/%s/unload", modelName),
+				"application/json",
+				bytes.NewBufferString("{}"))
+			if err != nil {
+				p.logger.Errorf("Failed to Unload model %s", modelName)
+				return
+			}
+			defer func(body io.ReadCloser) {
+				newErr := body.Close()
+				if newErr != nil {
+					p.logger.Error(newErr, "failed to close body")
+				}
+			}(resp.Body)
+			if resp.StatusCode == 200 {
+				p.logger.Infof("Successfully unloaded model %s", modelName)
+			} else {
+				body, err := io.ReadAll(resp.Body)
+				if err == nil {
+					p.logger.Infof("Failed to unload model %s with status [%d] and resp:%v", modelName, resp.StatusCode, body)
+				}
+			}
+		}
+		p.completions <- op
+
+	}
+
+	for op := range ops {
+		processOp(op)
 	}
 }

```

Reply:
@terrytangyuan nit:  another small update that might have been lost in the midst of that huge diff is using early returns in the switch statement, which would further reduce indentation and improve readability.

Reply:
Thanks. I just updated the code to return earlier, except for the return right before defer statement since otherwise defer won't be executed.

Reply:
Actually `return` also needs to be `break` instead since there's one more statement in that function that needs to be executed: `p.completions <- op`